Assets/PatchKit Patcher/Scripts/UI/DownloadSpeed.cs

```diff
@@ -15,27 +15,41 @@ public class DownloadSpeed : MonoBehaviour
 
         private void Start()
         {
-            Patcher.Instance.UpdateAppStatusChanged += status =>
+            var downloadStatus = Patcher.Instance.UpdaterStatus
+                .SelectSwitchOrNull(u => u.LatestActiveOperation)
+                .Select(s => s as IReadOnlyDownloadStatus);
+
+            var downloadSpeedUnit = Patcher.Instance.AppInfo.Select(a => a.PatcherDownloadSpeedUnit);
+
+            var text = downloadStatus.SelectSwitchOrDefault(status =>
             {
-                string speed = GetDownloadSpeedFormated(status);
-                string eta = status.DownloadBytes > 0 && status.DownloadBytes < status.DownloadTotalBytes && status.DownloadBytesPerSecond > 0.0
-                    ? " ready in " + GetRemainingTimeFormated(GetRemainingTime(status))
-                    : string.Empty;
+                var remainingTime =
+                    status.Bytes.CombineLatest<long, long, double, double?>(status.TotalBytes, status.BytesPerSecond,
+                        GetRemainingTime);
+
+                var formattedRemainingTime = remainingTime.Select<double?, string>(GetFormattedRemainingTime);
+
+                var formattedDownloadSpeed =
+                    status.BytesPerSecond.CombineLatest<double, string, string>(downloadSpeedUnit,
+                        GetFormattedDownloadSpeed);
 
-                Text.text = status.IsDownloading ? speed + eta : string.Empty;
-            };
+                return formattedDownloadSpeed.CombineLatest<string, string, string>(formattedRemainingTime,
+                    GetStatusText);
+            }, string.Empty);
 
-            Patcher.Instance.AppInfo.Subscribe(app => { _downloadSpeedUnit = app.PatcherDownloadSpeedUnit; })
-                .AddTo(this);
+            text.ObserveOnMainThread().SubscribeToText(Text).AddTo(this);
+        }
 
-            Text.text = string.Empty;
+        private static string GetStatusText(string formattedDownloadSpeed, string formattedRemainingTime)
+        {
+            return formattedDownloadSpeed + " " + formattedRemainingTime;
         }
 
-        private string GetDownloadSpeedFormated(OverallStatus status)
+        private static string GetFormattedDownloadSpeed(double bytesPerSecond, string downloadSpeedUnit)
         {
-            double kbPerSecond = status.DownloadBytesPerSecond / 1024.0;
+            double kbPerSecond = bytesPerSecond / 1024.0;
 
-            switch (_downloadSpeedUnit)
+            switch (downloadSpeedUnit)
             {
                 case "kilobytes":
```

Comment:
It would be good to keep these in constants somewhere related.

Reply:
I think it will be a good idea to add enums generator to PatchKit.Api.